pkg/configs/db/postgres/postgres.go

```diff
@@ -134,6 +134,97 @@ func (d DB) GetConfigs(since configs.ID) (map[string]configs.View, error) {
 	})
 }
 
+// GetRulesConfig gets the latest alertmanager config for a user.
+func (d DB) GetRulesConfig(userID string) (configs.VersionedRulesConfig, error) {
+	current, err := d.GetConfig(userID)
+	if err != nil {
+		return configs.VersionedRulesConfig{}, err
+	}
+	cfg := current.GetVersionedRulesConfig()
+	if cfg == nil {
+		return configs.VersionedRulesConfig{}, sql.ErrNoRows
+	}
+	return *cfg, nil
+}
+
+// SetRulesConfig sets the current alertmanager config for a user.
+func (d DB) SetRulesConfig(userID string, oldConfig, newConfig configs.RulesConfig) (bool, error) {
+	updated := false
+	err := d.Transaction(func(tx DB) error {
+		current, err := d.GetConfig(userID)
+		if err != nil && err != sql.ErrNoRows {
+			return err
+		}
+		// The supplied oldConfig must match the current config. If no config
+		// exists, then oldConfig must be nil. Otherwise, it must exactly
+		// equal the existing config.
+		if !((err == sql.ErrNoRows && oldConfig == nil) || oldConfig.Equal(current.Config.RulesFiles)) {
+			return nil
+		}
+		new := configs.Config{
+			AlertmanagerConfig: current.Config.AlertmanagerConfig,
+			RulesFiles:         newConfig,
+		}
+		updated = true
+		return d.SetConfig(userID, new)
+	})
+	return updated, err
+}
+
+// findRulesConfigs helps GetAllRulesConfigs and GetRulesConfigs retrieve the
+// set of all active rules configurations across all our users.
+func (d DB) findRulesConfigs(filter squirrel.Sqlizer) (map[string]configs.VersionedRulesConfig, error) {
+	rows, err := d.Select("id", "owner_id", "config ->> 'rules_files'").
```

Comment:
OK, now I don't understand why this doesn't bring back the entire history, so you will end up with some random version.

Reply:
ah, "`SELECT DISTINCT ON ( expression [, ...] )` keeps only the first row of each set of rows"
and you have order by `id DESC` below.

So effectively the same as `LIMIT 1` elsewhere

Reply:
Want me to add a comment along those lines? If so, can you share the link to the doc you just quoted?

Reply:
https://www.postgresql.org/docs/9.5/static/sql-select.html#SQL-DISTINCT